src/log.rs

```diff
@@ -0,0 +1,396 @@
+// Copyright 2020, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib_sys;
+use once_cell::sync::Lazy;
+use std::boxed::Box as Box_;
+use std::sync::Mutex;
+use translate::*;
+use GString;
+use LogLevelFlags;
+use LogWriterOutput;
+
+pub enum LogValue {
+    String(String),
+    Value(Vec<u8>),
+}
+
+pub struct LogField {
+    key: GString,
+    value: LogValue,
+}
+
+#[doc(hidden)]
+impl ToGlib for LogField {
+    type GlibType = glib_sys::GLogField;
+
+    fn to_glib(&self) -> glib_sys::GLogField {
+        match self.value {
+            LogValue::String(ref s) => {
+                let s: *const libc::c_char = s.to_glib_none().0;
+                glib_sys::GLogField {
+                    key: self.key.to_glib_none().0,
+                    value: s as _,
+                    length: -1,
+                }
+            }
+            LogValue::Value(ref v) => glib_sys::GLogField {
+                key: self.key.to_glib_none().0,
+                value: v.as_ptr() as _,
+                length: v.len() as _,
+            },
+        }
+    }
+}
+
+#[doc(hidden)]
+impl<'a> ToGlibPtr<'a, *const glib_sys::GLogField> for &'a [&LogField] {
+    type Storage = Vec<glib_sys::GLogField>;
+
+    #[inline]
+    fn to_glib_none(&'a self) -> Stash<'a, *const glib_sys::GLogField, Self> {
+        let fields = self.iter().map(|x| x.to_glib()).collect::<Vec<_>>();
+        Stash(fields.as_ptr() as *const _, fields)
+    }
+
+    #[inline]
+    fn to_glib_full(&self) -> *const glib_sys::GLogField {
+        let ptr = unsafe {
+            glib_sys::g_malloc0(::std::mem::size_of::<glib_sys::GLogField>() * self.len())
+                as *mut glib_sys::GLogField
+        };
+        for (pos, x) in self.iter().enumerate() {
+            unsafe {
+                *(ptr.offset(pos as _)) = x.to_glib();
+            }
+        }
+        ptr as *const _
+    }
+}
+
+#[cfg(any(feature = "v2_46", feature = "dox"))]
+pub fn log_set_handler_full<P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static>(
+    log_domain: Option<&str>,
+    log_levels: LogLevelFlags,
+    log_func: P,
+) -> u32 {
```

Comment:
I would make a custom type for the return value, just like `SourceId`. And then also use that for the removal